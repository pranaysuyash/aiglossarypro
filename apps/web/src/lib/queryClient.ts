import { isInLogoutState } from '@/lib/authPersistence';
import { MemoryMonitor } from '@/lib/MemoryMonitor';
import { IndexedDBManager } from '@/lib/IndexedDBManager';
import { QueryClient, type QueryFunction } from '@tanstack/react-query';

async function throwIfResNotOk(res: Response) {
  if (!res.ok) {
    const text = (await res.text()) || res.statusText;
    throw new Error(`${res.status}: ${text}`);
  }
}

export async function apiRequest(
  method: string,
  url: string,
  data?: unknown | undefined
): Promise<Response> {
  const requestInit: RequestInit = {
    method,
    headers: data ? { 'Content-Type': 'application/json' } : {},
    credentials: 'include',
  };

  if (data) {
    requestInit.body = JSON.stringify(data);
  }

  const res = await fetch(url, requestInit);

  await throwIfResNotOk(res);
  return res;
}

type UnauthorizedBehavior = 'returnNull' | 'throw';
export const getQueryFn: <T>(options: { on401: UnauthorizedBehavior }) => QueryFunction<T> =
  ({ on401: unauthorizedBehavior }) =>
    async ({ queryKey }) => {
      // Check if we're in logout state for auth queries
      if (queryKey[0] === '/api/auth/user' && isInLogoutState()) {
        console.log('ðŸš« Blocking auth query during logout state');
        return null;
      }

      // Prepare headers
      const headers: Record<string, string> = {};

      // Add Authorization header if token exists
      const token = localStorage.getItem('authToken');
      if (token) {
        headers.Authorization = `Bearer ${token}`;
      }

      const res = await fetch(queryKey[0] as string, {
        credentials: 'include',
        headers,
      });

      if (unauthorizedBehavior === 'returnNull' && res.status === 401) {
        return null;
      }

      await throwIfResNotOk(res);
      const jsonResponse = await res.json();

      // Extract data from API response format {success: true, data: ...}
      if (
        jsonResponse &&
        typeof jsonResponse === 'object' &&
        'success' in jsonResponse &&
        'data' in jsonResponse
      ) {
        return jsonResponse.data;
      }

      // Return full response if not in expected format
      return jsonResponse;
    };

// Memory-optimized query client with automatic cleanup
export const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      queryFn: getQueryFn({ on401: 'returnNull' }),
      refetchInterval: false,
      refetchOnWindowFocus: false,
      refetchOnReconnect: false,
      staleTime: 300000, // 5 minutes
      gcTime: 600000, // 10 minutes (formerly cacheTime)
      retry: (failureCount, error) => {
        // Don't retry on 4xx errors except 408, 429
        if (error instanceof Error && error.message.includes('4')) {
          const status = parseInt(error.message.split(':')[0]);
          if (status >= 400 && status < 500 && status !== 408 && status !== 429) {
            return false;
          }
        }
        // Retry up to 2 times for other errors
        return failureCount < 2;
      },
      retryDelay: (attemptIndex) => Math.min(1000 * 2 ** attemptIndex, 30000),
    },
    mutations: {
      retry: (failureCount, error) => {
        // Don't retry mutations on client errors
        if (error instanceof Error && error.message.includes('4')) {
          const status = parseInt(error.message.split(':')[0]);
          if (status >= 400 && status < 500) {
            return false;
          }
        }
        // Retry once for server errors
        return failureCount < 1;
      },
      retryDelay: 1000,
    },
  },
});

// Memory optimization: Set cache size limits
const MAX_CACHE_SIZE = 50; // Maximum number of queries to cache
const MEMORY_CHECK_INTERVAL = 30000; // Check memory every 30 seconds

// Register cleanup function with memory monitor (only in browser environment)
if (typeof window !== 'undefined') {
  try {
    MemoryMonitor.registerCleanupFunction(() => {
      console.log('ðŸ§¹ Emergency query cache cleanup triggered');

      // Get all queries and sort by last access time
      const queryCache = queryClient.getQueryCache();
      const queries = queryCache.getAll();

      // Remove oldest queries if cache is too large
      if (queries.length > MAX_CACHE_SIZE) {
        const sortedQueries = queries
          .sort((a, b) => (a.state.dataUpdatedAt || 0) - (b.state.dataUpdatedAt || 0))
          .slice(0, queries.length - MAX_CACHE_SIZE);

        sortedQueries.forEach(query => {
          queryCache.remove(query);
        });

        console.log(`ðŸ§¹ Removed ${sortedQueries.length} old queries from cache`);
      }

      // Force garbage collection of query cache
      queryClient.getQueryCache().clear();
      queryClient.getMutationCache().clear();
    });
  } catch (error) {
    console.warn('Failed to register memory monitor cleanup:', error);
  }
}

// Periodic cache optimization
let cacheOptimizationInterval: NodeJS.Timeout | null = null;

function startCacheOptimization() {
  if (cacheOptimizationInterval) {
    return;
  }

  cacheOptimizationInterval = setInterval(() => {
    const queryCache = queryClient.getQueryCache();
    const queries = queryCache.getAll();

    // Check if cache size exceeds limit
    if (queries.length > MAX_CACHE_SIZE) {
      console.log(`ðŸ§¹ Cache size (${queries.length}) exceeds limit (${MAX_CACHE_SIZE}), cleaning up...`);

      // Remove stale queries first
      const staleQueries = queries.filter(query => {
        const staleTime = query.options.staleTime || 0;
        const dataUpdatedAt = query.state.dataUpdatedAt || 0;
        return Date.now() - dataUpdatedAt > staleTime;
      });

      staleQueries.forEach(query => {
        queryCache.remove(query);
      });

      // If still too many, remove oldest queries
      const remainingQueries = queryCache.getAll();
      if (remainingQueries.length > MAX_CACHE_SIZE) {
        const sortedQueries = remainingQueries
          .sort((a, b) => (a.state.dataUpdatedAt || 0) - (b.state.dataUpdatedAt || 0))
          .slice(0, remainingQueries.length - MAX_CACHE_SIZE);

        sortedQueries.forEach(query => {
          queryCache.remove(query);
        });
      }

      console.log(`ðŸ§¹ Cache cleanup completed, ${queryCache.getAll().length} queries remaining`);
    }

    // Check memory usage and trigger GC if needed
    const memoryStats = MemoryMonitor.getMemoryStats();
    if (memoryStats.current && memoryStats.current.usedJSHeapSize > 256 * 1024 * 1024) {
      MemoryMonitor.triggerGarbageCollection();
    }
  }, MEMORY_CHECK_INTERVAL);
}

function stopCacheOptimization() {
  if (cacheOptimizationInterval) {
    clearInterval(cacheOptimizationInterval);
    cacheOptimizationInterval = null;
  }
}

// Start cache optimization when module loads (only in browser)
if (typeof window !== 'undefined') {
  startCacheOptimization();

  // Cleanup on page unload
  window.addEventListener('beforeunload', () => {
    stopCacheOptimization();
  });
}

// Export cache management functions
export const cacheManager = {
  startOptimization: startCacheOptimization,
  stopOptimization: stopCacheOptimization,
  getCacheSize: () => queryClient.getQueryCache().getAll().length,
  clearCache: () => {
    queryClient.getQueryCache().clear();
    queryClient.getMutationCache().clear();
  },
  optimizeCache: () => {
    const queryCache = queryClient.getQueryCache();
    const queries = queryCache.getAll();

    if (queries.length > MAX_CACHE_SIZE) {
      const sortedQueries = queries
        .sort((a, b) => (a.state.dataUpdatedAt || 0) - (b.state.dataUpdatedAt || 0))
        .slice(0, queries.length - MAX_CACHE_SIZE);

      sortedQueries.forEach(query => {
        queryCache.remove(query);
      });

      return sortedQueries.length;
    }

    return 0;
  },
};
