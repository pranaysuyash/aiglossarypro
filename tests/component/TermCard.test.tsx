import React from 'react';
import { describe, it, expect, vi, beforeEach } from 'vitest';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { Router } from 'wouter';
import TermCard from '../../client/src/components/TermCard';
import { useAuth } from "@/hooks/useAuth";
import { useToast } from "@/hooks/use-toast";

// Mock the API module
vi.mock('../../client/src/lib/api', () => ({
  apiRequest: vi.fn(),
}));

import * as queryClientModule from '../../client/src/lib/queryClient';

// Mock term data
const mockTerm = {
  id: '1',
  name: 'Machine Learning',
  shortDefinition: 'A method of data analysis that automates analytical model building',
  definition: 'Machine Learning is a subset of artificial intelligence that provides systems the ability to automatically learn and improve from experience without being explicitly programmed.',
  category: 'AI/ML',
  subcategories: ['Supervised Learning', 'Unsupervised Learning'],
  viewCount: 142,
  isFavorite: false,
  isLearned: false,
};

// Test wrapper with QueryClient and Router
const TestWrapper = ({ children }: { children: React.ReactNode }) => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
      mutations: {
        retry: false,
      },
    },
  });
  
  return (
    <QueryClientProvider client={queryClient}>
      <Router>
        {children}
      </Router>
    </QueryClientProvider>
  );
};

describe('TermCard Component', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  describe('Basic Rendering', () => {
    it('renders term card with basic information', () => {
      render(
        <TestWrapper>
          <TermCard term={mockTerm} />
        </TestWrapper>
      );
      
      expect(screen.getByText('Machine Learning')).toBeInTheDocument();
      expect(screen.getByText(/A method of data analysis/)).toBeInTheDocument();
    });

    it('displays essential term information', () => {
      render(
        <TestWrapper>
          <TermCard term={mockTerm} />
        </TestWrapper>
      );
      
      // Check for core content that should be displayed
      expect(screen.getByText('Machine Learning')).toBeInTheDocument();
      expect(screen.getByText(/A method of data analysis/)).toBeInTheDocument();
    });

    it('shows category information', () => {
      render(
        <TestWrapper>
          <TermCard term={mockTerm} />
        </TestWrapper>
      );
      
      expect(screen.getByText('AI/ML')).toBeInTheDocument();
    });

    it('handles missing optional fields gracefully', () => {
      const minimalTerm = {
        id: '2',
        name: 'Minimal Term',
        definition: 'A minimal definition'
      };

      render(
        <TestWrapper>
          <TermCard term={minimalTerm} />
        </TestWrapper>
      );

      expect(screen.getByText('Minimal Term')).toBeInTheDocument();
    });
  });

  describe('Interactive Features', () => {
    it('renders favorite state correctly', () => {
      const favoriteTerm = { ...mockTerm, isFavorite: true };
      render(
        <TestWrapper>
          <TermCard term={favoriteTerm} />
        </TestWrapper>
      );
      
      // Check for favorite indicator (star icon or similar)
      const favoriteIndicator = screen.getByRole('button', { name: /favorite/i });
      expect(favoriteIndicator).toBeInTheDocument();
    });

    it('handles favorite toggle interaction', async () => {
      const onFavoriteToggle = vi.fn();
      render(
        <TestWrapper>
          <TermCard term={mockTerm} onFavoriteToggle={onFavoriteToggle} />
        </TestWrapper>
      );

      const favoriteButton = screen.getByRole('button', { name: /favorite/i });
      fireEvent.click(favoriteButton);

      await waitFor(() => {
        expect(onFavoriteToggle).toHaveBeenCalledWith(mockTerm.id, true);
      });
    });

    it('renders learned state correctly', () => {
      const learnedTerm = { ...mockTerm, isLearned: true };
      render(
        <TestWrapper>
          <TermCard term={learnedTerm} />
        </TestWrapper>
      );
      
      // Check for learned indicator
      const learnedIndicator = screen.getByRole('button', { name: /learned/i });
      expect(learnedIndicator).toBeInTheDocument();
    });

    it('should navigate to term details on click', () => {
      render(
        <TestWrapper>
          <TermCard term={mockTerm} />
        </TestWrapper>
      );

      const termLink = screen.getByRole('link');
      expect(termLink).toHaveAttribute('href', `/term/${mockTerm.id}`);
    });
  });

  describe('Error Handling', () => {
    it('handles favorite toggle errors gracefully', async () => {
      const apiRequestSpy = vi.spyOn(queryClientModule, 'apiRequest');
      apiRequestSpy.mockRejectedValue(new Error('Network error'));

      const consoleSpy = vi.spyOn(console, 'error').mockImplementation(() => {});

      render(
        <TestWrapper>
          <TermCard term={mockTerm} />
        </TestWrapper>
      );

      const favoriteButton = screen.getByRole('button', { name: /favorite/i });
      fireEvent.click(favoriteButton);

      await waitFor(() => {
        expect(consoleSpy).toHaveBeenCalled();
      });

      consoleSpy.mockRestore();
    });

    it('handles malformed term data', () => {
      const malformedTerm = {
        id: null,
        name: '',
        definition: null
      } as any;

      // Should not crash with malformed data
      expect(() => {
        render(
          <TestWrapper>
            <TermCard term={malformedTerm} />
          </TestWrapper>
        );
      }).not.toThrow();
    });
  });

  describe('Performance', () => {
    it('should render efficiently with large datasets', () => {
      const renderStart = performance.now();
      
      render(
        <TestWrapper>
          <TermCard term={mockTerm} />
        </TestWrapper>
      );
      
      const renderTime = performance.now() - renderStart;
      expect(renderTime).toBeLessThan(100); // Should render quickly
    });
  });

  describe('Visual Snapshots', () => {
    it('matches visual snapshot', () => {
      const { container } = render(
        <TestWrapper>
          <TermCard term={mockTerm} />
        </TestWrapper>
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches visual snapshot with favorite and learned states', () => {
      const stateTerm = { ...mockTerm, isFavorite: true, isLearned: true };
      const { container } = render(
        <TestWrapper>
          <TermCard term={stateTerm} />
        </TestWrapper>
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
