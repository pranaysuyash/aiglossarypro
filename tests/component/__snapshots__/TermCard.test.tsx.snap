// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TermCard Component > Visual Snapshots > matches visual snapshot 1`] = `
<div
  class="rounded-lg border bg-card text-card-foreground shadow-sm h-full flex flex-col transition-shadow hover:shadow-md"
>
  <div
    class="p-4 flex-1"
  >
    <div
      class="flex items-center justify-between mb-2"
    >
      <div
        class="inline-flex items-center rounded-full border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-blue-100 text-blue-700 hover:bg-blue-200 dark:bg-blue-900 dark:text-blue-300"
      >
        AI/ML
      </div>
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:bg-accent h-8 w-8 text-gray-400 hover:text-accent-500"
      >
        <svg
          class="lucide lucide-heart"
          fill="none"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 14c1.49-1.46 3-3.21 3-5.5A5.5 5.5 0 0 0 16.5 3c-1.76 0-3 .5-4.5 2-1.5-1.5-2.74-2-4.5-2A5.5 5.5 0 0 0 2 8.5c0 2.3 1.5 4.05 3 5.5l7 7Z"
          />
        </svg>
      </button>
    </div>
    <h3
      class="font-semibold text-lg mb-1"
    >
      Machine Learning
    </h3>
    <p
      class="text-gray-600 dark:text-gray-400 text-sm mb-3 line-clamp-3"
    >
      A method of data analysis that automates analytical model building
    </p>
    <div
      class="text-xs text-gray-500 dark:text-gray-400 mb-2"
    >
      <span
        class="font-medium"
      >
        Categories: 
      </span>
      <span>
        Supervised Learning, Unsupervised Learning
      </span>
    </div>
  </div>
  <div
    class="border-t border-gray-100 dark:border-gray-800 p-3 flex justify-between items-center"
  >
    <div
      class="text-primary-600 hover:text-primary-700 dark:text-primary-400 dark:hover:text-primary-300 text-sm font-medium cursor-pointer"
    >
      Read more
    </div>
    <div
      class="flex space-x-2"
    >
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:bg-accent hover:text-accent-foreground h-8 w-8"
        data-state="closed"
      >
        <svg
          class="lucide lucide-copy text-gray-500 dark:text-gray-400"
          fill="none"
          height="16"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="14"
            rx="2"
            ry="2"
            width="14"
            x="8"
            y="8"
          />
          <path
            d="M4 16c-1.1 0-2-.9-2-2V4c0-1.1.9-2 2-2h10c1.1 0 2 .9 2 2"
          />
        </svg>
      </button>
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:bg-accent hover:text-accent-foreground h-8 w-8"
        data-state="closed"
      >
        <svg
          class="lucide lucide-share2 text-gray-500 dark:text-gray-400"
          fill="none"
          height="16"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="18"
            cy="5"
            r="3"
          />
          <circle
            cx="6"
            cy="12"
            r="3"
          />
          <circle
            cx="18"
            cy="19"
            r="3"
          />
          <line
            x1="8.59"
            x2="15.42"
            y1="13.51"
            y2="17.49"
          />
          <line
            x1="15.41"
            x2="8.59"
            y1="6.51"
            y2="10.49"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`TermCard Component > Visual Snapshots > matches visual snapshot with favorite and learned states 1`] = `
<div
  class="rounded-lg border bg-card text-card-foreground shadow-sm h-full flex flex-col transition-shadow hover:shadow-md"
>
  <div
    class="p-4 flex-1"
  >
    <div
      class="flex items-center justify-between mb-2"
    >
      <div
        class="inline-flex items-center rounded-full border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-blue-100 text-blue-700 hover:bg-blue-200 dark:bg-blue-900 dark:text-blue-300"
      >
        AI/ML
      </div>
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:bg-accent h-8 w-8 text-gray-400 hover:text-accent-500"
      >
        <svg
          class="lucide lucide-heart"
          fill="none"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 14c1.49-1.46 3-3.21 3-5.5A5.5 5.5 0 0 0 16.5 3c-1.76 0-3 .5-4.5 2-1.5-1.5-2.74-2-4.5-2A5.5 5.5 0 0 0 2 8.5c0 2.3 1.5 4.05 3 5.5l7 7Z"
          />
        </svg>
      </button>
    </div>
    <h3
      class="font-semibold text-lg mb-1"
    >
      Machine Learning
    </h3>
    <p
      class="text-gray-600 dark:text-gray-400 text-sm mb-3 line-clamp-3"
    >
      A method of data analysis that automates analytical model building
    </p>
    <div
      class="text-xs text-gray-500 dark:text-gray-400 mb-2"
    >
      <span
        class="font-medium"
      >
        Categories: 
      </span>
      <span>
        Supervised Learning, Unsupervised Learning
      </span>
    </div>
  </div>
  <div
    class="border-t border-gray-100 dark:border-gray-800 p-3 flex justify-between items-center"
  >
    <div
      class="text-primary-600 hover:text-primary-700 dark:text-primary-400 dark:hover:text-primary-300 text-sm font-medium cursor-pointer"
    >
      Read more
    </div>
    <div
      class="flex space-x-2"
    >
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:bg-accent hover:text-accent-foreground h-8 w-8"
        data-state="closed"
      >
        <svg
          class="lucide lucide-copy text-gray-500 dark:text-gray-400"
          fill="none"
          height="16"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="14"
            rx="2"
            ry="2"
            width="14"
            x="8"
            y="8"
          />
          <path
            d="M4 16c-1.1 0-2-.9-2-2V4c0-1.1.9-2 2-2h10c1.1 0 2 .9 2 2"
          />
        </svg>
      </button>
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:bg-accent hover:text-accent-foreground h-8 w-8"
        data-state="closed"
      >
        <svg
          class="lucide lucide-share2 text-gray-500 dark:text-gray-400"
          fill="none"
          height="16"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="18"
            cy="5"
            r="3"
          />
          <circle
            cx="6"
            cy="12"
            r="3"
          />
          <circle
            cx="18"
            cy="19"
            r="3"
          />
          <line
            x1="8.59"
            x2="15.42"
            y1="13.51"
            y2="17.49"
          />
          <line
            x1="15.41"
            x2="8.59"
            y1="6.51"
            y2="10.49"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`TermCard Component > matches visual snapshot 1`] = `
<div
  class="rounded-lg border bg-card text-card-foreground shadow-sm h-full flex flex-col transition-shadow hover:shadow-md"
>
  <div
    class="p-4 flex-1"
  >
    <div
      class="flex items-center justify-between mb-2"
    >
      <div
        class="inline-flex items-center rounded-full border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-blue-100 text-blue-700 hover:bg-blue-200 dark:bg-blue-900 dark:text-blue-300"
      >
        AI/ML
      </div>
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:bg-accent h-8 w-8 text-gray-400 hover:text-accent-500"
      >
        <svg
          class="lucide lucide-heart"
          fill="none"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 14c1.49-1.46 3-3.21 3-5.5A5.5 5.5 0 0 0 16.5 3c-1.76 0-3 .5-4.5 2-1.5-1.5-2.74-2-4.5-2A5.5 5.5 0 0 0 2 8.5c0 2.3 1.5 4.05 3 5.5l7 7Z"
          />
        </svg>
      </button>
    </div>
    <h3
      class="font-semibold text-lg mb-1"
    >
      Machine Learning
    </h3>
    <p
      class="text-gray-600 dark:text-gray-400 text-sm mb-3 line-clamp-3"
    >
      A method of data analysis that automates analytical model building
    </p>
    <div
      class="text-xs text-gray-500 dark:text-gray-400 mb-2"
    >
      <span
        class="font-medium"
      >
        Categories: 
      </span>
      <span>
        Supervised Learning, Unsupervised Learning
      </span>
    </div>
  </div>
  <div
    class="border-t border-gray-100 dark:border-gray-800 p-3 flex justify-between items-center"
  >
    <div
      class="text-primary-600 hover:text-primary-700 dark:text-primary-400 dark:hover:text-primary-300 text-sm font-medium cursor-pointer"
    >
      Read more
    </div>
    <div
      class="flex space-x-2"
    >
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:bg-accent hover:text-accent-foreground h-8 w-8"
        data-state="closed"
      >
        <svg
          class="lucide lucide-copy text-gray-500 dark:text-gray-400"
          fill="none"
          height="16"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="14"
            rx="2"
            ry="2"
            width="14"
            x="8"
            y="8"
          />
          <path
            d="M4 16c-1.1 0-2-.9-2-2V4c0-1.1.9-2 2-2h10c1.1 0 2 .9 2 2"
          />
        </svg>
      </button>
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:bg-accent hover:text-accent-foreground h-8 w-8"
        data-state="closed"
      >
        <svg
          class="lucide lucide-share2 text-gray-500 dark:text-gray-400"
          fill="none"
          height="16"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="18"
            cy="5"
            r="3"
          />
          <circle
            cx="6"
            cy="12"
            r="3"
          />
          <circle
            cx="18"
            cy="19"
            r="3"
          />
          <line
            x1="8.59"
            x2="15.42"
            y1="13.51"
            y2="17.49"
          />
          <line
            x1="15.41"
            x2="8.59"
            y1="6.51"
            y2="10.49"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`TermCard Component > matches visual snapshot with favorite and learned states 1`] = `
<div
  class="rounded-lg border bg-card text-card-foreground shadow-sm h-full flex flex-col transition-shadow hover:shadow-md"
>
  <div
    class="p-4 flex-1"
  >
    <div
      class="flex items-center justify-between mb-2"
    >
      <div
        class="inline-flex items-center rounded-full border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-blue-100 text-blue-700 hover:bg-blue-200 dark:bg-blue-900 dark:text-blue-300"
      >
        AI/ML
      </div>
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:bg-accent h-8 w-8 text-gray-400 hover:text-accent-500"
      >
        <svg
          class="lucide lucide-heart"
          fill="none"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 14c1.49-1.46 3-3.21 3-5.5A5.5 5.5 0 0 0 16.5 3c-1.76 0-3 .5-4.5 2-1.5-1.5-2.74-2-4.5-2A5.5 5.5 0 0 0 2 8.5c0 2.3 1.5 4.05 3 5.5l7 7Z"
          />
        </svg>
      </button>
    </div>
    <h3
      class="font-semibold text-lg mb-1"
    >
      Machine Learning
    </h3>
    <p
      class="text-gray-600 dark:text-gray-400 text-sm mb-3 line-clamp-3"
    >
      A method of data analysis that automates analytical model building
    </p>
    <div
      class="text-xs text-gray-500 dark:text-gray-400 mb-2"
    >
      <span
        class="font-medium"
      >
        Categories: 
      </span>
      <span>
        Supervised Learning, Unsupervised Learning
      </span>
    </div>
  </div>
  <div
    class="border-t border-gray-100 dark:border-gray-800 p-3 flex justify-between items-center"
  >
    <div
      class="text-primary-600 hover:text-primary-700 dark:text-primary-400 dark:hover:text-primary-300 text-sm font-medium cursor-pointer"
    >
      Read more
    </div>
    <div
      class="flex space-x-2"
    >
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:bg-accent hover:text-accent-foreground h-8 w-8"
        data-state="closed"
      >
        <svg
          class="lucide lucide-copy text-gray-500 dark:text-gray-400"
          fill="none"
          height="16"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="14"
            rx="2"
            ry="2"
            width="14"
            x="8"
            y="8"
          />
          <path
            d="M4 16c-1.1 0-2-.9-2-2V4c0-1.1.9-2 2-2h10c1.1 0 2 .9 2 2"
          />
        </svg>
      </button>
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:bg-accent hover:text-accent-foreground h-8 w-8"
        data-state="closed"
      >
        <svg
          class="lucide lucide-share2 text-gray-500 dark:text-gray-400"
          fill="none"
          height="16"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="18"
            cy="5"
            r="3"
          />
          <circle
            cx="6"
            cy="12"
            r="3"
          />
          <circle
            cx="18"
            cy="19"
            r="3"
          />
          <line
            x1="8.59"
            x2="15.42"
            y1="13.51"
            y2="17.49"
          />
          <line
            x1="15.41"
            x2="8.59"
            y1="6.51"
            y2="10.49"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
